blackjax/mcmc/langevin.py

```diff
@@ -0,0 +1,43 @@
+"""Solvers for Langevin diffusions."""
+import jax
+import jax.numpy as jnp
+
+from blackjax.types import PyTree, PRNGKey
+from blackjax.base import SamplingAlgorithm
+
+from typing import Callable, NamedTuple, Tuple
+from blackjax.types import Array, PRNGKey, PyTree
+from blackjax.mcmc.gradients import GradientState
+
+__all__ = ["overdamped_langevin"]
+
+class DiffusionState(NamedTuple):
+    position: PyTree
+    grad_state: GradientState
+
+
+def generate_noise(rng_key: PRNGKey, position):
+    position_flat, unravel_fn = jax.flatten_util.ravel_pytree(position)
+    noise_flat = jax.random.normal(rng_key, shape=jnp.shape(position_flat))
+    noise = unravel_fn(noise_flat)
+    return noise
+
+def overdamped_langevin(logprob_grad_fn):
+    """
+    Euler solver for overdamped Langevin diffusion
+    """
+    def one_step(rng_key, state: DiffusionState, step_size: float, batch: tuple=()):
+        position, grad_state = state
+        noise = generate_noise(rng_key, position)
+        position = jax.tree_util.tree_multimap(
+            lambda p, g, n: p + step_size * g + jnp.sqrt(2 * step_size) * n,
+            position,
+            grad_state.logprob_grad,
+            noise,
+        )
+
+        new_state = DiffusionState(position, grad_state)
+        grad_state = GradientState(logprob_grad_fn(new_state, *batch))
```

Comment:
We could pass `jax.grad(logprob_fn)` as `logprob_grad_fn` and call `logprob_grad_fn(position, *batch)`

Reply:
I pass in `DiffusionState` rather than position because SVRG will need the current centering value to estimate the gradient. This will be in `grad_state: GradientState` in `DiffusionState`.

If we pass in only the position here then the diffusion can't be re-used.
What do you think? Is there perhaps another solution?